docs/proposals/20220330-topology-mutation-hook.md

```diff
@@ -0,0 +1,359 @@
+* Explore a solution to detect and prevent an External Patch Extension to trigger infinite reconciles
+
+
+## Proposal
+
+### User Stories
+
+As ClusterClass author:
+* I want to enable an External Patch Extension which injects an HTTP proxy configuration into my Cluster topology.
+* I want to enable an External Patch Extension which computes Machine images for my Cluster topology given the Kubernetes version and the region.
+* I want to enable an External Patch Extension which enables image pulls from a private registry by injecting certificates into the Machines of my Cluster topology.
+* I want to easily reuse an External Patch Extension in many ClusterClasses.
+
+As an External Patch Extension developer:
+* I want to use my preferred programming language to implement my External Patch Extension.
+* I want to unit test the code/logic which generates external patches.
+* I want to be able to generate external patches in either JSON Patch or JSON Merge Patch format.
+* I want to generate external patches based on external data, for example by querying a cloud API.
+* I want to validate the templates after all patches have been applied, so I can be sure that other External Patch Extensions didn't overwrite my changes.
+
+### Cluster Operator guide
+
+As a Cluster operator, to use ClusterClasses with an External Patch Extensions you have to deploy and register it. You can find the full documentation on how to deploy a Runtime Extension in the [Runtime SDK proposal](https://github.com/kubernetes-sigs/cluster-api/blob/75b39db545ae439f4f6203b5e07496d3b0a6aa75/docs/proposals/20220221-runtime-SDK.md#deploy-runtime-extensions).
+
+An External Patch Extension can be registered by applying:
+```yaml
+apiVersion: runtime.cluster.x-k8s.io/v1beta1
+kind: Extension
+metadata:
+  name: "my-awesome-patch"
+spec:
+  clientConfig:
+    service:
+      namespace: "capi-extensions"
+      name: "my-awesome-patch"
+```
+
+Once the extension is registered the discovery hook is called and the Extension CR is updated with the list of the Runtime Extensions supported by the server. Most notably each Runtime Extension will get a unique name assigned, which is then used in the ClusterClass. For more details on discovery please see the [Runtime SDK proposal](https://github.com/kubernetes-sigs/cluster-api/blob/9a3a8f557db850b4a67e9e8e0a70cb2d0261caee/docs/proposals/20220221-runtime-SDK.md#registering-runtime-extensions).
+
+### ClusterClass author guide
+
+A ClusterClass author can use an External Patch Extension by referencing it in a ClusterClass and adding the corresponding variable definitions.
+
+A ClusterClass can have external patches, inline patches or both. The patches will then be applied in the order in which they are defined. The `hook` field of the external patch must match the unique name of a RuntimeExtension assigned during discovery.
```

Comment:
Just for my understanding. This means that an `Extension` must always be deployed before a `ClusterClass` that references it, because the computed name is unkown beforehand and will only be visible in the `status` after the `Extension` has been registered successfully?

Also why is the field named `hook` when it actually references an `extension`? Shouldn't this reference the name field like mentioned here: https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220221-runtime-SDK.md?plain=1#L559
By the above sentence I would think of using the name of the field here instead: https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220221-runtime-SDK.md?plain=1#L562
But probably I'm just conflating the terms hook and extension here :sweat_smile:

Reply:
> This means that an Extension must always be deployed before a ClusterClass that references it, because the computed name is unkown beforehand and will only be visible in the status after the Extension has been registered successfully?

AFAIK as I remember computed name value is predictable, it is the concatenation of the name returned by discovery and the name the Extension object 
@sbueringer to confirm

> Also why is the field named hook when it actually references an extension? 

This makes sense to me, +1 to disambiguate if possible

Reply:
> AFAIK as I remember computed name value is predictable, it is the concatenation of the name returned by discovery and the name the Extension object

That is correct. If you use a well-defined and documented External Patch Extension the name is deterministic. Otherwise deploying concurrently is not an issue as we retry the Cluster reconciliation until the the corresponding `Extension` exists and discovery is done.

> Also why is the field named hook when it actually references an extension?

Agree. The current name is not ideal. I would wait for the upcoming Extension CRD PR though as I think we have to think about how we call "Discovered RuntimeExtensions" there and this might lead to an obvious name here.

Reply:
Just FYI: The Extension CRD PR is now merged. Naming the field `extensionHandler` will be in sync with the naming used in the Extension CRD. Open to other suggestions.

Reply:
I'm not sure if we just call it extensionHandler that the external patching aspect is clear enough.
I see the following variations / dimensions:
1. do we want to the fields top-level or nested (like we did it with inline patches with the "definition" field)
2. how do we name the fields

One idea:
```yaml
  patches:
  - name: external-patch-1
    external:
      generate:
      validate:
  - name: config
    definitions:
    - selector:
    ....
```

Alternatives:
1. inline the fields instead of having them in `external`
2. Alternative field variations:
* generate+validate
* generatePatches+validateTopology
* mutate+validate
* generationHandler+validationHandler
* generatePatchesHandler+validateTopologyHandler
* mutationHandler+validationHandler
* Similar with "Extension": generationExtension+validationExtension, ...


@fabriziopandini @killianmuldoon @vincepri @ykakarap Opinions?

Reply:
I would vote for having nested fields under `external`. As for the field names how about `generator` (generates the patch) and `validator` (validates the topology)?

Names from the above list that are also okay:
* generatePatches+validateTopology
* generationHandler+validationHandler

Let's not use mutate because it doesn't really mutate the object it only send back the patch to be applied. Using mutate could have a different impression.

Reply:
+1 to nest with generateExtension+validateExtension